### PR TITLE
Different/configurable distance metrics

### PR DESCRIPTION
**Pull request by wdm0006**
## Background

We've been doing some work with LIME in relatively high-dimension problems (1000+ features), and have found that the current distance metric doesn't tend to perform well in that context [[related paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.23.7409&rep=rep1&type=pdf)].  For very high dimensions, the euclidian distance is inflated to a large value, which given the kernel you are using:

`sqrt(exp(-d^2/k^2))`

Can be tuned by using very large kernel widths.  A simple example is the classification problem defined by using a default sklearn RandomForestClassifier on a dataset created with:

`sklearn.datasets.make_classification(n_classes=2, n_features=1000, n_samples=1000)`

In this case, a kernel width of 3 leads to the weight for the first record (the sample being interpreted itself) having a weight of 1, while all other samples are weighted at near 0, which of course leads to a bad interpretation.  Using very large kernel_widths increases the average score (~0.3 with a width of 10e9), but in general, I think this just isn't the best metric for higher dimension problems.  I've tried some other metrics, and without tuning the kernel width, they are performing much better than euclidian norm can do with a well tuned kernel width).

Just for reference, using the default kernel width of 3, and changing only distance metric on the above problem led to a range of scores between 0 and 1, with cityblock and L1 distance performing best of those I tried, at the cost of a small slowdown.  The scores per sample within each run had low variance, but the scores per metric differed quite a bit, which indicates to me that this is an important tuning parameter.  Chebyshev distance performed better than euclidean but not the best, and was slightly faster than the current euclidean code.

## Changes

This PR exposes an option to the user to specify a metric from this pretty comprehensive list: [sklearn pairwise distances](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.pairwise_distances.html), defaulting to euclidian for tabular and cosine for text which is the the same as the current measures.  So for existing users, this changes nothing, but optionally, for cases like I described above, you'll be able to use different metrics.

I've also included a benchmarks folder which has a script for each of the explainers that loops through a sample dataset and records scores and times, it may be useful for other similar optimizations.

### CODE REVIEW

**Comment by marcotcr** [comment]
This is a problem that I had already identified and was on my 'todo' stack. The problem with euclidean distance for tabular is not only that it is bad for high dimensions, but that it is really hard to come up with a sensible default value (as you noted yourself).

The benchmark folder is also a good idea.

Thanks you very much, I will merge this soon.

**Comment by wdm0006** [comment]
Cool, happy to help. 

Perhaps it would be useful to have a contributing.md file with some of this kind of 'todo' item for existing or other potential contributors?

**Comment by marcotcr** [comment]
Yes, that's a great idea. When I have a bit of time, I'll think try to write a file with known problems and ideas.

If you have any intuition or suggestions for a reasonable distance / kernel pair for tabular data, let me know! 

Thanks again for contributing.

**Comment by wdm0006** [comment]
I'll put together a PR with some experimental results but I think just scaling the distance by the number of columns before squaring it makes a single default kernel width work well with a wider range of datasets.  I've only tested it with the Euclidean distance so far but I would expect pretty similar improvements or no change for the other metrics.
